### PR TITLE
Add tensorboard support.

### PR DESCRIPTION
**Pull request by afiaka87**
Weights and biases is cool - but they don't really support the offline usecase as well as tensorboard does currently. This fix simply adds tensorboard as a fallback option - W&B will still run if it's available.

### CODE REVIEW

**Comment by lucidrains** [comment]
oh interesting, does tensorboard work here where weights and biases does not?

**Comment by afiaka87** [comment]
> oh interesting, does tensorboard work here where weights and biases does not?

Helmholtz distributed nodes makes this difficult as GPU nodes don't have internet. They do however support Tensorboard tunnelled over ssh. Wandb uses docker to run the local version which is a bundle of issues I don't want to dive into _quite_ yet. The only solution is to run Wandb offline, and then sync the run after training. If you're using DeepSpeed or horovod while it's training, you will corrupt your Wandb cache preventing upload - there's an issue on their GitHub about it that I'll have to find. 

Also - we've been having that index error on the call to `wandb.Image` when running with fp16. This doesn't fix that but DeepSpeed _does_ have Tensorboard support built in which may fix the issue.

Basically the network effect from Tensorboard being so popular generally is quite helpful for these otherwise tough to solve problems with logging distributed. I obviously enjoy and still use W&B myself because you get checkpoint versions and better default graphs, reports, etc. But this is useful for anyone who isn't interested in/can't run wandb. At least to be able visualize results on JUWELS. Debugging the training itself without live reconstructions makes it easy to miss bugs where the loss is fine but the reconstructions are still bad, for instance.

It does add yet another cli arg. I could maybe split it out into it's own script if you'd prefer?

**Comment by janEbert** [comment]
> If you're using DeepSpeed or horovod while it's training, you will corrupt your Wandb cache preventing upload - there's an issue on their GitHub about it that I'll have to find but my recollection that it was largely people on distributed compute with similar such restrictions.

Unless WandB does some evil reflection shenanigans with the `globals()` dictionary, whether training is distributed should not have an effect on its data getting corrupted. I'm assuming people with issues do not properly guard their WandB functions so they are only executed on a single worker – parallel writes will always cause problems.

> Also - we've been having that index error on the call to wandb.Image when running with fp16. This doesn't fix that but DeepSpeed does have Tensorboard support built in which may fix the issue.

About this, the error is not related to WandB in any way, but it is related to DeepSpeed. When `DALLE.generate_images` is called, the first iteration [in the for-loop](https://github.com/lucidrains/DALLE-pytorch/blob/ef44c98c3fbfe1a00c7ae7ae9350f8ad0e6e4818/dalle_pytorch/dalle_pytorch.py#L415-L430) correctly executes. However, the second time around, [the call to `torch.topk`](https://github.com/lucidrains/DALLE-pytorch/blob/ef44c98c3fbfe1a00c7ae7ae9350f8ad0e6e4818/dalle_pytorch/dalle_pytorch.py#L48) in [`top_k` (called here)](https://github.com/lucidrains/DALLE-pytorch/blob/ef44c98c3fbfe1a00c7ae7ae9350f8ad0e6e4818/dalle_pytorch/dalle_pytorch.py#L422) fails. It's really weird and I haven't been able to debug why this happens yet.
My [`deepspeed` branch](https://github.com/janEbert/DALLE-pytorch/tree/deepspeed) has more proper handling for this with regard to partitioned DeepSpeed models but I still get the same error.

**Comment by AyushExel** [comment]
Hey @afiaka87. A member W&B team here. Agreed with @janEbert . Most of the DDP problems arise because they're executed on multiple workers as opposed to the controller process( rank 0). If this doesn't work as expected, then it's a bug. If you already found the github issue just point me to that and I'll get someone to look into it.

**Comment by afiaka87** [comment]
> Hey @afiaka87. A member W&B team here. Agreed with @janEbert . Most of the DDP problems arise because they're executed on multiple workers as opposed to the controller process( rank 0). If this doesn't work as expected, then it's a bug. If you already found the github issue just point me to that and I'll get someone to look into it.

This one in particular is very common for me @AyushExel - I'll follow up later today. Thanks for clarifying that W&B isn't necessarily at issue here @janEbert - is the following quote at least not still true for either of you? @lucidrains @janEbert 

> tl;dr Tensorboard is a bit more established still and wandb is blocking us on a few things that tensorboard wouldn't necessarily.

**Comment by afiaka87** [comment]
@AyushExel This issue is very similar to mine. There are a few others in the Issues section which are quite similar actually but the recurring trend is the same - someone tried to sync their local `wandb` runs _during_ training, and it broke everything. This has happened to me quite often.

I remember crossing paths with an older issue by others which referenced mpirun. Unfortunately I don't have time to find that one specifically. 

https://github.com/wandb/client/issues/1647

Love W&B by the way! I use it every day.

**Comment by afiaka87** [comment]
@lucidrains @janEbert 
In any case - I hope it's clear that I never tried to say `wandb` was broken for distributed. I simply said:

(emphasis mine)
> Also - we've been having that index error on the call to wandb.Image when running with fp16. This doesn't fix that but DeepSpeed does have Tensorboard support built in which **may fix the issue**

I was wrong about that.

My biggest pain point is not actually wandb itself, but rather the inability to have a live feedback loop of results _during_ training. As @lucidrains can attest to - sometimes the loss can look fine but the re-constructions can be broken for subtle reasons. Spending 2 hours of compute to be able to make that discovery is my only issue here - not necessarily W&B (although, I will just say in terms of velocity - tensorboard works on julich today without hassle. W&B does not. This is not a bug but rather an administrative/ops issue.

@lucidrains I'm taking it to mean that you'd prefer we kept things lean with W&B? This code is useful to me personally but I admit it only exists to help me on a contrived setup which most users don't deal with (julich). I'm going to try to prioritize the indexing issues instead.

@AyushExel - do follow up with me on that bug. 

The workflow is basically this (just to give you more detail):

User signs into login node that has internet acesss.
User installs wandb with `pip install wandb --user`
user run `wandb offline` to make sure there aren't timeouts/etc. while running on the GPU node.
User submits task to GPU node - GPU node runs in the background.
_on the login node_ the user types 

`wandb online`
`wandb sync --sync-all`

And due (presumably) to a race condition of some sort, wandb will then place a corrupted file of some sort in your runs folder. Now, any attempt to run `wandb sync` will result in failure.

**Comment by afiaka87** [comment]
> Unless WandB does some evil reflection shenanigans with the `globals()` dictionary, whether training is distributed should not have an effect on its data getting corrupted. I'm assuming people with issues do not properly guard their WandB functions so they are only executed on a single worker – parallel writes will always cause problems.

@janEbert - have you tried using `wandb` in offline mode and then syncing the run as it trains? Because that is rather broken for me.

**Comment by AyushExel** [comment]
> @AyushExel This issue is very similar to mine. There are a few others in the Issues section which are quite similar actually but the recurring trend is the same - someone tried to sync their local `wandb` runs _during_ training (after having enabled `wandb offline` and then re-enabling it after training began), and it corruped a file and makes it difficult/impossible to upload. This has happened to me quite often.
> 
> I remember crossing paths with an older issue by others which referenced mpirun. Unfortunately I don't have time to find that one specifically.
> 
> [wandb/client#1647](https://github.com/wandb/client/issues/1647)
> 
> Love W&B by the way! I use it every day.

@afiaka87 the thread suggests that this issue was not reproducible and the cli version is 0.10.12 which is much older than the current stable release( 0.10.27). Nevertheless, I'll bring this issue up with the CLI team to see if we can improve it. But I'm glad you're able to use tensorboard to accomplish what you need :)

**Comment by janEbert** [comment]
> @janEbert - have you tried using wandb in offline mode and then syncing the run as it trains? Because that is rather broken for me.

I always disabled it for simplicity, sorry. Can't help with this. :(

**Comment by afiaka87** [comment]
> > @AyushExel This issue is very similar to mine. There are a few others in the Issues section which are quite similar actually but the recurring trend is the same - someone tried to sync their local `wandb` runs _during_ training (after having enabled `wandb offline` and then re-enabling it after training began), and it corruped a file and makes it difficult/impossible to upload. This has happened to me quite often.
> > I remember crossing paths with an older issue by others which referenced mpirun. Unfortunately I don't have time to find that one specifically.
> > [wandb/client#1647](https://github.com/wandb/client/issues/1647)
> > Love W&B by the way! I use it every day.
> 
> @afiaka87 the thread suggests that this issue was not reproducible and the cli version is 0.10.12 which is much older than the current stable release( 0.10.27). Nevertheless, I'll bring this issue up with the CLI team to see if we can improve it. But I'm glad you're able to use tensorboard to accomplish what you need :)

If I could reproduce the issue on the latest version of `wandb` would the team then be able to prioritize it better?
@AyushExel

**Comment by AyushExel** [comment]
@afiaka87 yes for sure. Also, if you could export the wandb logs from that run that crashed while syncing, it'll be much easier to get to the source of the problem. Meanwhile, I'll ask someone to take a look at this thread to see it this points to any known issue

**Comment by afiaka87** [comment]
> > @janEbert - have you tried using wandb in offline mode and then syncing the run as it trains? Because that is rather broken for me.
> 
> I always disabled it for simplicity

Kinda the crux of my argument.

**Comment by afiaka87** [comment]
> @afiaka87 yes for sure. Also, if you could export the wandb logs from that run that crashed while syncing, it'll be much easier to get to the source of the problem. Meanwhile, I'll ask someone to take a look at this thread to see if this points to any known issue

Thanks! I'll get back to you about this on the `wandb` issues later today hopefully!

**Comment by lucidrains** [comment]
@afiaka87 I'd like to keep W&B's, but if the support engineer cannot work through the issues with you, then I will have no choice but to accept alternatives

**Comment by lucidrains** [comment]
@afiaka87 ok sounds good! just want to give them a chance to fix the blocking issues - we can consider tensorboard if and only if the obstacles are insurmountable

**Comment by afiaka87** [comment]
> @afiaka87 ok sounds good! just want to give them a chance to fix the blocking issues - we can consider tensorboard if and only if the obstacles are insurmountable

Yeah no worries - that's one of the reasons I made this a draft.

**Comment by afiaka87** [comment]
@AyushExel 

I made an extensive bug report including a script to reproduce, a video of the bug, and plenty of environment info. I was ultimately able to sync my report after I stopped training, but what we are looking for is the ability to run a _manual_ sync as needed.

https://github.com/wandb/client/issues/2132

**Comment by afiaka87** [comment]
@lucidrains it seems they're aware of the issue but it's not currently a high priority. 

https://github.com/wandb/client/issues/2132

The given solution is good enough for me I guess but they _did_ just go down https://status.wandb.com/ and having a backup in such cases would be nice.

**Comment by mehdidc** [comment]
+1 for this PR, I had similar problems with wandb, I think support for both would be useful if there is no way wandb could work in a complete offline way.

**Comment by mehdidc** [comment]
@afiaka87 Minor fix: the creation of the writer should be done in rank zero `if distr_backend.is_root_worker()` to avoid having one tensorboard file per rank, except if that was intended

**Comment by robvanvolt** [comment]
Does the tensorboard support work well? Or is there a reason it is still not merged to main?:)

**Comment by afiaka87** [comment]
Uhh I kinda lost motivation on getting this to work properly _around_ the same time I was finally able to get this running on my local RTX2070 and no longer needed to use the distributed setup (which was at issue - W&B breaks on JULICH bc their GPU nodes cant access the internet). 

Also I think there might be a simpler way to do this - W&B has built in tensorboard support that just automatically logs to both wandb.ai and to a local/remote tensorboard URL. Probably the better way to do it. Anyway - I think I'll close this for now? It's pretty out of date anyways and wouldn't be fun to merge/rebase and contains plenty of flaws as well (like not matching the captions to your images in tensorboard/not using the built-in W&B tensorboard plugin).

@robvanvolt lemme know if you're having issues with W&B; I remember being a bit disgruntled with them at the time of this PR because they had just had an outage which broke training for me and had a culture on their github repo of not really caring at all about the distributed/non-cloud scenario. I think they kinda turned around a bit after that because of backlash but I don't know honestly.
